src/tiledb/cloud/vcf/ingestion.py

```diff
@@ -0,0 +1,977 @@
+        create_log_array(log_uri)
+        group = tiledb.Group(dataset_uri, "w")
+        group.add(log_uri, name=LOG_ARRAY)
+        group.close()
+
+        write_log_event(log_uri, "create_dataset_udf", "create", data=dataset_uri)
+
+        create_manifest(dataset_uri)
+    else:
+        logger.info(f"Using existing dataset: '{dataset_uri}'.")
+
+    return dataset_uri
+
+
+def read_uris_udf(
+    dataset_uri: str,
+    list_uri: str,
+    *,
+    config: Optional[Mapping[str, Any]] = None,
+    max_files: Optional[int] = None,
+) -> Sequence[str]:
+    """
+    Read a list of URIs from a URI.
+
+    :param dataset_uri: dataset URI
+    :param list_uri: URI of the list of URIs
+    :param config: config dictionary, defaults to None
+    :param max_files: maximum number of URIs returned, defaults to None
+    :return: list of URIs
+    """
+
+    logger, profiler = setup(config, dataset_uri)
+
+    result = []
+    vfs = tiledb.VFS()
+    for line in vfs.open(list_uri):
+        result.append(line.decode().strip())
+        if max_files and len(result) == max_files:
+            break
+
+    logger.info(f"Found {len(result)} VCF files.")
+    profiler.finish()
+
+    return result
+
+
+def find_uris_udf(
+    dataset_uri: str,
+    search_uri: str,
+    *,
+    config: Optional[Mapping[str, Any]] = None,
+    pattern: Optional[str] = None,
+    ignore: Optional[str] = None,
+    max_files: Optional[int] = None,
+) -> Sequence[str]:
+    """
+    Find URIs matching a pattern in a directory or S3 bucket using the following command:
+
+        find <search_uri> -name <pattern> | grep -Ev <ignore> -m <max_files>
+
+    with an efficient implementation for S3.
+
+    :param dataset_uri: dataset URI
+    :param search_uri: URI to search for VCF files
+    :param config: config dictionary, defaults to None
+    :param pattern: pattern used in the search, defaults to None
+    :param ignore: exclude pattern applied to the search results, defaults to None
+    :param max_files: maximum number of URIs returned, defaults to None
+    :return: list of URIs
+    """
+
+    logger, profiler = setup(config, dataset_uri)
+
+    use_s3 = search_uri.startswith("s3://")
+    # Run command to find URIs matching the pattern
+    if use_s3:
+        cmd = f"exec aws s3 sync --dryrun --exclude '*' --include '{pattern}' {search_uri} ."
+    else:
+        cmd = f"find {search_uri} -name {pattern}"
+    logger.debug(cmd)
+    p1 = subprocess.Popen(
+        cmd,
+        shell=True,
+        text=True,
+        stdout=subprocess.PIPE,
+    )
+
+    # Optionally ignore URIs and limit the number returned
+    if ignore:
+        cmd = f"grep -Ev '{ignore}'"
+    else:
+        cmd = "grep ''"
+    if max_files:
+        cmd += f" -m {max_files}"
+
+    logger.debug(cmd)
+    p2 = subprocess.Popen(
+        cmd,
+        shell=True,
+        text=True,
+        stdin=p1.stdout,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+    )
```

Comment:
Could we do this in Python, iteratively reading in lines and handling them as needed rather than shelling out to grep?